server/src/thirdparty/slackAPI/service.js

```diff
@@ -0,0 +1,162 @@
+import axios from 'axios';
+import env from '../env';
+import config from './config';
+import ThirdPartyApp from '../models/ThirdPartyApp';
+import tpEventEmitter, { generateActionUpdateEvent, generateAppAddedEvent, generateAppUpdatedEvent } from '../event';
+
+const getAuthUrl = (code) => `${config.api}/api/oauth.access?client_id=${env.slackClientId}&client_secret=${env.slackClientSecret}&code=${code}&redirect_uri=${encodeURIComponent(`${env.baseUri}/tp/slack/auth/redirect`)}`;
+
+const getIntegUrl = (state) => `${config.api}/oauth/authorize?scope=bot,channels:history,channels:read,chat:write:bot&client_id=${env.slackClientId}&redirect_uri=${encodeURIComponent(`${env.baseUri}/tp/slack/auth/redirect`)}&state=${state}`;
+
+const generateAuthToken = (code) => {
+  const authPromise = new Promise((resolve, reject) => {
+    axios.get(getAuthUrl(code)).then((response) => {
+      resolve(response.data);
+    }).catch((error) => {
+      reject(error);
+    });
+  });
+  return authPromise;
+};
+
+const getIntegrationId = async (userId, teamId) => {
+  const result = await ThirdPartyApp
+    .findOne({
+      'userDetails.team_id': teamId,
+      'userDetails.user_id': userId,
+    })
+    .select({ _id: 1 }).exec();
+
+  return result === null ? null : result._id;
+};
+
+const getToken = async (integrationId) => {
+  const { token } = await ThirdPartyApp
+    .findOne({
+      _id: integrationId,
+    })
+    .select({ token: 1, _id: 0 })
+    .exec();
+
+  return token;
+};
+
+const generateNewEvent = async (teamId, authedUsers, event) => {
+  const eventJson = { eventData: event };
+  eventJson.appName = config.appName;
+  eventJson.event = `${event.type}_${event.channel_type}`;
+  ThirdPartyApp
+    .find({
+      'userDetails.team_id': teamId,
+      'userDetails.user_id': { $in: authedUsers },
+    })
+    .select({ _id: 1 })
+    .exec((err, apps) => {
+      eventJson.integratedApps = apps.map((app) => app._id);
+      if (apps.length !== 0) {
+        tpEventEmitter.emit('newEvent', eventJson);
+      }
+    });
+};
+
+const addOrUpdateIntegration = async (authData, userToken) => {
+  if (authData.ok === false) {
+    throw new Error('Access token generation failed');
+  }
+  // eslint-disable-next-line camelcase
+  const { team_name, access_token, user_id, team_id } = authData;
+  const integrationId = await getIntegrationId(user_id, team_id);
+  const slackApp = {
+    appName: config.appName,
+    accountName: team_name,
+    token: access_token,
+    userDetails: { user_id, team_id },
+  };
+  let status = false;
```

Comment:
instead of using `let` you can call a function and return the values.
like
```
const somethingHappening = () => {
  if (true) return 1;
  return 2;
};
const result = somethingHappening();
```

Reply:
Done